transports/janus_mqtt.c

```diff
@@ -0,0 +1,648 @@
+			janus_mqtt_client_message_arrived,
+			janus_mqtt_client_delivery_complete) != MQTTASYNC_SUCCESS) {
+		JANUS_LOG(LOG_FATAL, "Can't connect to MQTT broker: error setting up callbacks...\n");
+		goto error;
+	}
+
+	/* Connecting to the broker */
+	int rc = janus_mqtt_client_connect(ctx);
+	if(rc != MQTTASYNC_SUCCESS) {
+		JANUS_LOG(LOG_FATAL, "Can't connect to MQTT broker, return code: %d\n", rc);
+		goto error;
+	}
+
+	return 0;
+
+error:
+	/* If we got here, something went wrong */
+	janus_mqtt_client_destroy_context(ctx);
+	g_free((char *)url);
+	g_free((char *)client_id);
+	g_free((char *)username);
+	g_free((char *)password);
+	g_free(config);
+
+	return -1;
+}
+
+void janus_mqtt_destroy(void) {
+	JANUS_LOG(LOG_INFO, "Disconnecting MQTT client...\n");
+
+	janus_mqtt_client_disconnect(context_);
+}
+
+int janus_mqtt_get_api_compatibility(void) {
+	return JANUS_TRANSPORT_API_VERSION;
+}
+
+int janus_mqtt_get_version(void) {
+	return JANUS_MQTT_VERSION;
+}
+
+const char *janus_mqtt_get_version_string(void) {
+	return JANUS_MQTT_VERSION_STRING;
+}
+
+const char *janus_mqtt_get_description(void) {
+	return JANUS_MQTT_DESCRIPTION;
+}
+
+const char *janus_mqtt_get_name(void) {
+	return JANUS_MQTT_NAME;
+}
+
+const char *janus_mqtt_get_author(void) {
+	return JANUS_MQTT_AUTHOR;
+}
+
+const char *janus_mqtt_get_package(void) {
+	return JANUS_MQTT_PACKAGE;
+}
+
+gboolean janus_mqtt_is_janus_api_enabled(void) {
+	return janus_mqtt_api_enabled_;
+}
+
+gboolean janus_mqtt_is_admin_api_enabled(void) {
+	return janus_mqtt_admin_api_enabled_;
+}
+
+int janus_mqtt_send_message(void *context, void *request_id, gboolean admin, json_t *message) {
+	if(message == NULL) {
+		return -1;
+	}
+	if(context == NULL) {
+		json_decref(message);
+		return -1;
+	}
+
+	janus_mqtt_context *ctx = (janus_mqtt_context *)context;
+	char *payload = json_dumps(message, json_format_);
+	json_decref(message);
+
+	JANUS_LOG(LOG_HUGE, "Sending %s API message via MQTT: %s\n", admin ? "admin" : "Janus", payload);
+
+	if(admin) {
+		int rc = janus_mqtt_client_publish_admin_message(ctx, payload);
+		if(rc != MQTTASYNC_SUCCESS) {
+			JANUS_LOG(LOG_ERR, "Can't publish to MQTT topic: %s, return code: %d\n", ctx->admin.publish.topic, rc);
+			return -1;
+		}
+	}
+	else {
+		int rc = janus_mqtt_client_publish_janus_message(ctx, payload);
+		if(rc != MQTTASYNC_SUCCESS) {
+			JANUS_LOG(LOG_ERR, "Can't publish to MQTT topic: %s, return code: %d\n", ctx->publish.topic, rc);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+void janus_mqtt_session_created(void *context, guint64 session_id) {
+	/* We don't care */
+}
+
+void janus_mqtt_session_over(void *context, guint64 session_id, gboolean timeout) {
+	/* We don't care, not even if it's a timeout (should we?), our client is always up */
+}
+
+void janus_mqtt_client_connection_lost(void *context, char *cause) {
+	JANUS_LOG(LOG_INFO, "MQTT connection lost cause of %s. Reconnecting...\n", cause);
+	// automatic reconnect
```

Comment:
Shouldn't this notify the core the transport is gone, and then notify it again when it comes back up? I guess the idea is not to loose existing sessions because of an occasional disconnection, but a missing connection would result in notifications/events sent by the core being just dropped I believe.